### PR TITLE
[core] Add `line-gradient` property

### PR DESCRIPTION
**Pull request by pozdnyakov**
Porting of https://github.com/mapbox/mapbox-gl-js/pull/6303
See the link above for the description of the feature and
its limitations).

Based on patch from @lbud (Lauren Budorick).

### CODE REVIEW

**Comment by pozdnyakov** [comment]
This PR fixes https://github.com/mapbox/mapbox-gl-native/issues/11718

File: platform/darwin/src/MGLLineStyleLayer.h [hunk 1]
```diff
@@ -366,6 +366,50 @@ MGL_EXPORT
 */
 @property (nonatomic) MGLTransition lineGapWidthTransition;
 
+#if TARGET_OS_IPHONE
+/**
+ Defines a gradient with which to color a line feature. Can only be used with
+ GeoJSON sources that specify `"lineMetrics": true`.
```
Comment (1ec5) [inline platform/darwin/src/MGLLineStyleLayer.h]: We’ll need to customize this documentation comment by adding en entry to platform/darwin/scripts/style-spec-overrides-v8.json:

> This property only has an effect on lines defined by an `MGLShapeSource` whose `MGLShapeSource.calculatesLineDistanceMetrics` property is set to `YES`.

Speaking of which, we’ll need to add that `calculatesLineDistanceMetrics` property to MGLShapeSource before developers will be able to use this property on iOS/macOS.

File: platform/darwin/src/MGLLineStyleLayer.h [hunk 2]
```diff
@@ -366,6 +366,50 @@ MGL_EXPORT
...
+ 
+ This property is only applied to the style if `lineDasharray` is set to `nil`,
```
Comment (1ec5) [inline platform/darwin/src/MGLLineStyleLayer.h]: This should say `lineDashPattern`. The codegen script isn’t aliasing `lineDasharray` for some reason.

**Comment by pozdnyakov** [comment]
@1ec5 Thanks for your comments! I've made a https://github.com/mapbox/mapbox-gl-native/pull/12604 introducing a new property to `MGLShapeSource`, please take a look! After https://github.com/mapbox/mapbox-gl-native/pull/12604 lands I'll include the updated `style-spec-overrides-v8.json`  to this PR.

File: src/mbgl/renderer/buckets/line_bucket.cpp [hunk 3]
```diff
@@ -89,6 +105,25 @@ void LineBucket::addGeometry(const GeometryCoordinates& coordinates, const Geome
...
+    auto clip_start_it = props.find("mapbox_clip_start");
+    auto clip_end_it = props.find("mapbox_clip_end");
+    if (clip_start_it != props.end() && clip_end_it != props.end()) {
+        double total_length = [first, &coordinates, len] {
+            double result = 0.0;
+            for (std::size_t i = first; i < len - 1; ++i) {
+                result += util::dist<double>(coordinates[i], coordinates[i + 1]);
+            }
+            return result;
+        }();
```
Comment (kkaefer) [inline src/mbgl/renderer/buckets/line_bucket.cpp]: Why is this a self-invoking lambda?

File: src/mbgl/renderer/buckets/line_bucket.cpp [hunk 4]
```diff
@@ -392,14 +427,18 @@ void LineBucket::addGeometry(const GeometryCoordinates& coordinates, const Geome
...
                                   double endLeft,
                                   double endRight,
                                   bool round,
                                   std::size_t startVertex,
-                                  std::vector<TriangleElement>& triangleStore) {
+                                  std::vector<TriangleElement>& triangleStore,
+                                  optional<Distances> lineDistances) {
     Point<double> extrude = normal;
+    if (lineDistances)
+        distance = lineDistances->scaleToMaxLineDistance(distance);
```
Comment (kkaefer) [inline src/mbgl/renderer/buckets/line_bucket.cpp]: Let's add braces here even if there is just one statement.

File: src/mbgl/renderer/buckets/line_bucket.cpp [hunk 5]
```diff
@@ -392,14 +427,18 @@ void LineBucket::addGeometry(const GeometryCoordinates& coordinates, const Geome
 }
 
 void LineBucket::addCurrentVertex(const GeometryCoordinate& currentCoordinate,
-                                  double &distance,
+                                  double distance,
```
Comment (kkaefer) [inline src/mbgl/renderer/buckets/line_bucket.cpp]: Why do we no longer need the distance as a reference? We still seem to be modifying it.

File: src/mbgl/renderer/buckets/line_bucket.cpp [hunk 6]
```diff
@@ -436,8 +475,12 @@ void LineBucket::addPieSliceVertex(const GeometryCoordinate& currentVertex,
                                    const Point<double>& extrude,
                                    bool lineTurnsLeft,
                                    std::size_t startVertex,
-                                   std::vector<TriangleElement>& triangleStore) {
+                                   std::vector<TriangleElement>& triangleStore,
+                                   optional<Distances> lineDistances) {
     Point<double> flippedExtrude = extrude * (lineTurnsLeft ? -1.0 : 1.0);
+    if (lineDistances)
+        distance = lineDistances->scaleToMaxLineDistance(distance);
```
Comment (kkaefer) [inline src/mbgl/renderer/buckets/line_bucket.cpp]: Braces please :)

File: src/mbgl/style/expression/parsing_context.cpp [hunk 7]
```diff
@@ -68,7 +68,8 @@ bool isConstant(const Expression& expression) {
     }
     
     return isFeatureConstant(expression) &&
-        isGlobalPropertyConstant(expression, std::array<std::string, 2>{{"zoom", "heatmap-density"}});
+        isGlobalPropertyConstant(expression, std::array<std::string, 2>{{"zoom", "heatmap-density"}}) &&
+        isGlobalPropertyConstant(expression, std::array<std::string, 2>{{"zoom", "line-progress"}});;
```
Comment (kkaefer) [inline src/mbgl/style/expression/parsing_context.cpp]: Nit: double semicoli

File: src/mbgl/style/layers/line_layer.cpp [hunk 8]
```diff
@@ -431,5 +431,32 @@ TransitionOptions LineLayer::getLinePatternTransition() const {
...
+    return { {} };
+}
+
+ColorRampPropertyValue LineLayer::getLineGradient() const {
+    return impl().paint.template get<LineGradient>().value;
+}
+
+void LineLayer::setLineGradient(ColorRampPropertyValue value) {
+    if (value == getLineGradient())
+        return;
```
Comment (kkaefer) [inline src/mbgl/style/layers/line_layer.cpp]: Braces

File: src/mbgl/renderer/buckets/line_bucket.cpp (hunk 5 above)
Comment (pozdnyakov) [reply src/mbgl/renderer/buckets/line_bucket.cpp]: Thanks for the catch! I missed that we need to set it to `0.0` under some conditions

File: src/mbgl/style/layers/line_layer.cpp (hunk 8 above)
Comment (pozdnyakov) [reply src/mbgl/style/layers/line_layer.cpp]: This is a generated file, I could introduce braces generation to `scripts/generate-style-code.js` in a follow-up PR

**Comment by pozdnyakov** [comment]
@kkaefer could you take a look?

File: platform/darwin/src/MGLLineStyleLayer.h (hunk 1 above)
Comment (1ec5) [reply platform/darwin/src/MGLLineStyleLayer.h]: #13000

File: platform/darwin/src/MGLLineStyleLayer.h (hunk 2 above)
Comment (1ec5) [reply platform/darwin/src/MGLLineStyleLayer.h]: #13000

**Comment by friedbunny** [comment]
For future spelunkers: this was included in `release-frappé` (which was `ios-v4.4.0` and `android-v6.5.0`).

**Pull request merged**